board/safety/safety_volkswagen.h

```diff
@@ -1,142 +1,232 @@
+  relay_malfunction = false;
+  volkswagen_torque_msg = MSG_HCA_01;
+  volkswagen_lane_msg = MSG_LDW_02;
+  gen_crc_lookup_table(0x2F, volkswagen_crc8_lut_8h2f);
 
-    if (controls_allowed) {
+  return;
+}
 
-      // *** global torque limit check ***
-      violation |= max_limit_check(desired_torque, VOLKSWAGEN_MAX_STEER, -VOLKSWAGEN_MAX_STEER);
+static int volkswagen_mqb_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
 
-      // *** torque rate limit check ***
-      violation |= driver_limit_check(desired_torque, volkswagen_desired_torque_last, &volkswagen_torque_driver,
-        VOLKSWAGEN_MAX_STEER, VOLKSWAGEN_MAX_RATE_UP, VOLKSWAGEN_MAX_RATE_DOWN,
-        VOLKSWAGEN_DRIVER_TORQUE_ALLOWANCE, VOLKSWAGEN_DRIVER_TORQUE_FACTOR);
-      volkswagen_desired_torque_last = desired_torque;
+  bool valid = addr_safety_check(to_push, volkswagen_mqb_rx_checks, VOLKSWAGEN_MQB_RX_CHECKS_LEN,
+                                 volkswagen_get_checksum, volkswagen_mqb_compute_crc, volkswagen_get_counter);
 
-      // *** torque real time rate limit check ***
-      violation |= rt_rate_limit_check(desired_torque, volkswagen_rt_torque_last, VOLKSWAGEN_MAX_RT_DELTA);
+  if (valid) {
+    int bus = GET_BUS(to_push);
+    int addr = GET_ADDR(to_push);
+
+    // Update in-motion state by sampling front wheel speeds
+    // Signal: ESP_19.ESP_VL_Radgeschw_02 (front left)
+    // Signal: ESP_19.ESP_VR_Radgeschw_02 (front right)
+    if ((bus == 0) && (addr == MSG_ESP_19)) {
+      int wheel_speed_fl = GET_BYTE(to_push, 4) | (GET_BYTE(to_push, 5) << 8);
+      int wheel_speed_fr = GET_BYTE(to_push, 6) | (GET_BYTE(to_push, 7) << 8);
+      if ((wheel_speed_fl > 0) || (wheel_speed_fr > 0)) {
+        volkswagen_moving = true;
+      } else {
+        volkswagen_moving = false;
+      }
+    }
 
-      // every RT_INTERVAL set the new limits
-      uint32_t ts_elapsed = get_ts_elapsed(ts, volkswagen_ts_last);
-      if (ts_elapsed > VOLKSWAGEN_RT_INTERVAL) {
-        volkswagen_rt_torque_last = desired_torque;
-        volkswagen_ts_last = ts;
+    // Update driver input torque samples
+    // Signal: EPS_01.Driver_Strain (absolute torque)
+    // Signal: EPS_01.Driver_Strain_VZ (direction)
+    if ((bus == 0) && (addr == MSG_EPS_01)) {
+      int torque_driver_new = GET_BYTE(to_push, 5) | ((GET_BYTE(to_push, 6) & 0x1F) << 8);
+      int sign = (GET_BYTE(to_push, 6) & 0x80) >> 7;
+      if (sign == 1) {
+        torque_driver_new *= -1;
       }
+      update_sample(&volkswagen_torque_driver, torque_driver_new);
     }
 
-    // no torque if controls is not allowed
-    if (!controls_allowed && (desired_torque != 0)) {
-      violation = true;
+    // Update ACC status from drivetrain coordinator for controls-allowed state
+    // Signal: TSK_06.TSK_Status
+    if ((bus == 0) && (addr == MSG_TSK_06)) {
+      int acc_status = (GET_BYTE(to_push, 3) & 0x7);
+      controls_allowed = ((acc_status == 3) || (acc_status == 4) || (acc_status == 5)) ? 1 : 0;
```

Comment:
Replaced signal ACC_06.ACC_Status_ACC (from the ACC radar) with signal TSK_06.TSK_Status (from the TreibStraing-Koordinator, drivetrain coordinator). TSK_06 has the exact same info, and it will always be seen on Panda bus 0, whereas ACC_06 could be on bus 0 or 2 depending on the install, which is hard to accommodate in the post-ISO26262 era.

This will get a corresponding PR on the OP side shortly. TSK_06 will be a frame or so behind ACC_06. I've seen that if OP applies torque instantly on engagement based on ACC_06, Panda tx_hook can start clipping and create a temp steering fault just at the moment of engagement.